src/Client.php

```diff
@@ -275,4 +275,151 @@ public function getAsset($assetId) {
     return (string) $response->getBody();
   }
 
+  /**
+   * Gets presigned url from AWS S3.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam.
+   *
+   * @return mixed
+   *   Presigned url needed for next step + PID.
+   */
+  public function getPresignUrl(array $file_data) {
+    $this->checkAuth();
+
+    try {
+      $response = $this->client->request(
+        "GET",
+        $this->baseUrl . '/ws/awss3/generateupload',
+        [
+          'headers' => $this->getDefaultHeaders(),
+          'query' => $file_data,
+        ]
+      );
+      return json_decode($response->getBody());
+    }
+    catch (ClientException $e) {
+      dump($e->getMessage());
+      return FALSE;
+    }
+  }
+
+  /**
+   * Uploads file to Webdam AWS S3.
+   *
+   * @param mixed $presignedUrl
+   *   The presigned URL we got in previous step from AWS.
+   * @param array $file_data
+   *   The file data required by Webdam.
+   *
+   * @return array
+   *   Response Status 100 / 200
+   */
+  public function uploadPresigned($presignedUrl, array $file_data) {
+    $this->checkAuth();
+
+    $file = fopen($file_data['file_uri'], 'r');
+
+    try {
+      $response = $this->client->request(
+        "PUT",
+        $presignedUrl, [
+          'headers' => ['Content-Type' => $file_data['contenttype']],
+          'body' => stream_get_contents($file),
+        ]);
+
+      return [
+        'status' => json_decode($response->getStatusCode(), TRUE),
+      ];
+    }
+    catch (ClientException $e) {
+      dump($e->getMessage());
+      return [
+        'step' => 'upload presigned',
+        'error' => json_decode($e, TRUE),
+      ];
+    }
+  }
+
+  /**
+   * Confirms the upload to Webdam.
+   *
+   * @param string $pid
+   *   The Process ID we got in first step.
+   *
+   * @return array
+   *   The step, the status code and the newly uploaded asset ID.
+   */
+  public function uploadConfirmed($pid) {
+    $this->checkAuth();
+
+    try {
+      $response = $this->client->request(
+        "PUT",
+        $this->baseUrl . '/ws/awss3/finishupload/' . $pid,
+        ['headers' => $this->getDefaultHeaders()]
+      );
+
+      return [
+        'step' => 'upload confirmed',
+        'status' => $response->getStatusCode(),
+        'body' => json_decode($response->getBody(), TRUE)['id'],
+      ];
+    }
+    catch (ClientException $e) {
+      dump($e->getMessage());
+      return [
+        'step' => 'confirmation',
+        'error' => json_decode($e, TRUE),
+      ];
+    }
+  }
+
+  /**
+   * Uploads Assets to Webdam using the previously defined methods.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam.
+   * @param int $folderID
+   *   The Webdam folder ID.
+   *
+   * @return array
+   *   Webdam response.
+   */
+  public function uploadAsset(array $file_data, $folderID = NULL) {
```

Comment:
This is probably the only thing that needs to be public. Just pass in the contents of the file, a folder ID, and a filename to upload to, and then handle the three step process all in this function. You can call out to other methods in the class, but there's probably no reason to make those methods public - callers just want to upload the asset. They shouldn't have to care about specifically how that happens.